spinn_front_end_common/interface/interface_functions/command_sender_adder.py

```diff
@@ -0,0 +1,111 @@
+
+    def add_command_senders(self):
+        progress = ProgressBar(FecDataView.get_n_vertices(), "Adding commands")
+        for vertex in progress.over(FecDataView.iterate_vertices()):
+            if isinstance(vertex, AbstractSendMeMulticastCommandsVertex):
+                machine = FecDataView.get_machine()
+                link_data = None
+
+                # See if we need a specific placement for a device
+                if isinstance(vertex, ApplicationFPGAVertex):
+                    fpga = vertex.outgoing_fpga_connection
+                    if fpga is not None:
+                        link_data = machine.get_fpga_link_with_id(
+                            fpga.fpga_id, fpga.fpga_link_id,
+                            fpga.board_address, fpga.chip_coords)
+                elif isinstance(vertex, ApplicationSpiNNakerLinkVertex):
+                    link_data = machine.get_spinnaker_link_with_id(
+                        vertex.spinnaker_link_id, vertex.board_address)
+
+                command_sender = self.__get_command_sender(link_data)
+
+                # allow the command sender to create key to partition map
+                command_sender.add_commands(
+                    vertex.start_resume_commands,
+                    vertex.pause_stop_commands,
+                    vertex.timed_commands, vertex)
+
+        all_command_senders = list(self.__command_sender_for_chip.values())
+        if self.__general_command_sender is not None:
+            all_command_senders.append(self.__general_command_sender)
+
+        # add the edges from the command senders to the dependent vertices
+        for command_sender in all_command_senders:
+            FecDataView.add_vertex(command_sender)
+            edges, partition_ids = command_sender.edges_and_partitions()
+            for edge, partition_id in zip(edges, partition_ids):
+                FecDataView.add_edge(edge, partition_id)
+
+    def __cores(self, x, y):
+        return [p.processor_id
+                for p in FecDataView.get_chip_at(x, y).processors
+                if not p.is_monitor]
+
+    def __get_command_sender(self, link_data):
+        if link_data is None:
+            if self.__general_command_sender is None:
+                self.__general_command_sender = self.__new_command_sender()
+            return self.__general_command_sender
+
+        x = link_data.connected_chip_x
+        y = link_data.connected_chip_y
+
+        command_sender = self.__command_sender_for_chip.get((x, y))
+        if command_sender is None:
+            command_sender = self.__new_command_sender()
+            self.__command_sender_for_chip[(x, y)] = command_sender
+            cores = self.__cores(x, y)
+            p = cores[self.__system_placements.n_placements_on_chip(x, y)]
+            self.__system_placements.add_placement(
+                Placement(command_sender.machine_vertex, x, y, p))
+        return command_sender
+
+    def __new_command_sender(self):
```

Comment:
I would do this inline so you can add the xy to the label where applicable

Reply:
I added the label as an argument